multinet/api.py

```diff
@@ -41,14 +95,32 @@ def get_workspaces() -> Any:
     return stream
 
 
-@bp.route("/workspaces/<workspace>", methods=["GET"])
-@require_reader
-@swag_from("swagger/workspace.yaml")
-def get_workspace(workspace: str) -> Any:
-    """Retrieve a single workspace."""
+@bp.route("/workspaces/<workspace>/permissions", methods=["GET"])
+@require_maintainer
+@swag_from("swagger/get_workspace_permissions.yaml")
+def get_workspace_permissions(workspace: str) -> Any:
+    """Retrieve the permissions of a workspace."""
     metadata = db.get_workspace_metadata(workspace)
 
-    return {"name": metadata["name"], "permissions": metadata["permissions"]}
+    return _permissions_id_to_user(metadata["permissions"])
+
+
+@bp.route("/workspaces/<workspace>/permissions", methods=["PUT"])
+@require_maintainer
+@swag_from("swagger/set_workspace_permissions.yaml")
+def set_workspace_permissions(workspace: str) -> Any:
+    """Set the permissions on a workspace."""
+    if set(request.json.keys()) != {
+        "owner",
+        "maintainers",
+        "writers",
+        "readers",
+        "public",
+    }:
+        raise MalformedRequestBody(request.json)
```

Comment:
I think I might disagree with this. A `PUT` call usually means that you specify only what you want changed, and the server figures out how to apply your "diff".

This way of operating also works, but there's a danger: suppose you `GET` the permissions, then you have faulty logic for updating that object so that you accidentally put in changes you didn't want. Then on `PUT`, you will end up with bad state.

This is probably a weak argument. I'm willing to hear why it should be required to pass back the whole permissions object. One argument that feels at least somewhat compelling is that the server logic could become more complex to handle partial updates.

In any case, there's no need to change things in this PR. I'd like to have the discussion and if we decide that partial `PUT`s are warranted, we'll just file an issue and worry about it later.

Reply:
> I think I might disagree with this. A `PUT` call usually means that you specify only what you want changed, and the server figures out how to apply your "diff".

I disagree, I think of `PUT` as an updating an existing resource to match what you specify (more or less). You seem to be describing the `PATCH` method.


> This way of operating also works, but there's a danger: suppose you `GET` the permissions, then you have faulty logic for updating that object so that you accidentally put in changes you didn't want. Then on `PUT`, you will end up with bad state.

I think this could happen in any way of utilizing the endpoint. At the end of the day, it's our responsibility to make sure the client uses this endpoint properly. Plus, you still can't modify owners from these endpoints, so in the worst case the owner can still fix things after an error.

> This is probably a weak argument. I'm willing to hear why it should be required to pass back the whole permissions object. One argument that feels at least somewhat compelling is that the server logic could become more complex to handle partial updates.

Yeah this is one of the main reasons, and I don't really see a benefit to specifying only the change rather than the whole new object. It makes the logic on the server more complex, and it also makes the logic on the client much more complex.

Reply:
You are right, especially in pointing out that I'm thinking of `PATCH` rather than `PUT`. I agree with everything you said here.